src/main/java/org/medicmobile/webapp/mobile/SettingsDialogActivity.java

```diff
@@ -213,40 +225,85 @@ class ServerMetadata {
 		this.name = name;
 		this.url = url;
 	}
+
+	public String getName() {
+		return name;
+	}
 }
 
 class ServerRepo {
 	private final SharedPreferences prefs;
 
 	ServerRepo(Context ctx) {
 		prefs = ctx.getSharedPreferences(
-				"ServerRepo",
-				Context.MODE_PRIVATE);
-		save("https://gamma.dev.medicmobile.org");
-		save("https://gamma-cht.dev.medicmobile.org");
-		save("https://medic.github.io/atp");
+			"ServerRepo",
+			Context.MODE_PRIVATE);
+
+		Map<String, String> instances = parseInstanceXML(ctx);
+		for (Map.Entry<String, String> entry : instances.entrySet()) {
+			String instanceName = entry.getValue();
+			String instanceUrl = entry.getKey();
+
+			save(instanceName, instanceUrl);
+		}
 	}
 
-	List<ServerMetadata> getServers() {
+	List<ServerMetadata> getServers(Boolean allowCustomHosts) {
 		List servers = new LinkedList<ServerMetadata>();
 
-		servers.add(new ServerMetadata("Custom"));
-
 		for(Map.Entry<String, ?> e : prefs.getAll().entrySet()) {
 			servers.add(new ServerMetadata(
-					e.getValue().toString(),
-					e.getKey()));
+				e.getValue().toString(),
+				e.getKey()));
+		}
+
+		Collections.sort(servers, Comparator.comparing(ServerMetadata::getName));
+
+		if (allowCustomHosts) {
+			servers.add(0, new ServerMetadata("Custom"));
 		}
 
 		return servers;
 	}
 
-	void save(String url) {
+	void save(String name, String url) {
+		if (name == null) {
+			name = friendly(url);
+		}
+
 		SharedPreferences.Editor ed = prefs.edit();
-		ed.putString(url, friendly(url));
+		ed.putString(url, name);
 		ed.apply();
 	}
 
+	private static Map<String, String> parseInstanceXML(Context context) {
+		try {
+			HashMap<String, String> result = new HashMap<>();
```

Comment:
Isn't it great that Java does not have tuples!   :)  Android utils does have a `Pair` class, but FWIW I am not really sure it would improve much to make this `List<Pair<String, String>> result...`